tools/restatectl/src/connection.rs

```diff
@@ -0,0 +1,267 @@
+                Err(status) => {
+                    errors.error(address, status);
+                    continue;
+                }
+            };
+
+            let nodes_configuration =
+                StorageCodec::decode::<NodesConfiguration, _>(&mut response.encoded)
+                    .map_err(|err| ConnectionInfoError::InvalidNodesConfiguration(address, err))?;
+
+            if nodes_configuration.version()
+                > latest_configuration
+                    .as_ref()
+                    .map(|c| c.version())
+                    .unwrap_or(Version::INVALID)
+            {
+                latest_configuration = Some(nodes_configuration);
+            }
+        }
+
+        if !answer {
+            // all nodes has errored
+            return Err(ConnectionInfoError::NodesErrors(errors));
+        }
+
+        *config_guard = latest_configuration.clone();
+        latest_configuration.ok_or(ConnectionInfoError::MissingNodesConfiguration)
+    }
+
+    pub async fn try_each<F, T, Fut>(
+        &self,
+        role: Option<Role>,
+        mut closure: F,
+    ) -> Result<Response<T>, ConnectionInfoError>
+    where
+        F: FnMut(Channel) -> Fut,
+        Fut: Future<Output = Result<Response<T>, Status>>,
+    {
+        let nodes_config = self.get_nodes_configuration().await?;
+        let mut channels = self.cache.lock().await;
+
+        let iterator = match role {
+            Some(role) => Either::Left(nodes_config.iter_role(role)),
+            None => Either::Right(nodes_config.iter()),
+        }
+        .sorted_by(|left, right| {
+            // nodes that we already have channels open for, gets higher presence.
+            match (
+                channels.contains_key(&left.0),
+                channels.contains_key(&right.0),
+            ) {
+                (true, false) => Ordering::Less,
+                (false, true) => Ordering::Greater,
+                (_, _) => left.0.cmp(&right.0),
+            }
+        });
+
+        let mut errors = NodesErrors::default();
+
+        for (node_id, node) in iterator {
+            // avoid creating new channels on each iteration. Instead cheaply copy the channels
+            let channel = channels
+                .entry(node_id)
+                .or_insert_with(|| grpc_channel(node.address.clone()));
+
+            let result = closure(channel.clone()).await;
+            match result {
+                Ok(response) => return Ok(response),
+                Err(status) => {
+                    errors.error(node.address.clone(), status);
+                }
+            }
+        }
+
+        if errors.is_empty() {
+            Err(ConnectionInfoError::NoAvailableNodes(NoRoleError(role)))
+        } else {
+            Err(ConnectionInfoError::NodesErrors(errors))
+        }
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum ConnectionInfoError {
+    #[error("Could not retrieve nodes configuration. Possible un provisioned cluster!")]
```

Comment:
Unprovisioned sounds a bit awkward, maybe something like this is more consistent with the tone we take elsewhere?

```suggestion
    #[error("Could not retrieve nodes configuration. Have you provisioned this cluster?")]
```